### ISSUE TITLE
6X Coordinator crashes in MPPnoticeReceiver

### ISSUE THREAD

**Issue by Smyatkin-Maxim** [greenplum-db/gpdb#15194]
## Bug Report
I've got this stack tracewith a core file the other day:
```
#0  raise (sig=sig@entry=11) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x000055eaf83ee59f in StandardHandlerForSigillSigsegvSigbus_OnMainThread (processName=<optimized out>, postgres_signal_arg=11) at elog.c:5581
#2  <signal handler called>
#3  0x000055eaf87ccaec in strlcpy (dst=dst@entry=0x55eaf9864888 "", src=src@entry=0x0, siz=siz@entry=10) at strlcpy.c:56
#4  0x000055eaf846adb1 in MPPnoticeReceiver (arg=<optimized out>, res=<optimized out>) at cdbconn.c:716
#5  0x000055eaf81b4e38 in pqInternalNotice (hooks=hooks@entry=0x55eaf9932870, fmt=fmt@entry=0x55eaf88ea728 "message type 0x%02x arrived from server while idle") at fe-exec.c:893
#6  0x000055eaf81ab903 in pqParseInput3 (conn=0x55eaf9932780) at fe-protocol3.c:218
#7  0x000055eaf81b3a36 in parseInput (conn=conn@entry=0x55eaf9932780) at fe-exec.c:1746
#8  0x000055eaf81b6366 in PQisBusy (conn=0x55eaf9932780) at fe-exec.c:1763
#9  0x000055eaf846d0db in processResults (dispatchResult=dispatchResult@entry=0x55eaf99d1970) at cdbdisp_async.c:1000
#10 0x000055eaf846e0c8 in handlePollSuccess (fds=<optimized out>, pParms=<optimized out>) at cdbdisp_async.c:825
#11 checkDispatchResult (ds=0x55eaf99d1638, timeout_sec=-1) at cdbdisp_async.c:638
#12 0x000055eaf846bfbe in cdbdisp_checkDispatchResult (ds=ds@entry=0x55eaf99d1638, waitMode=waitMode@entry=DISPATCH_WAIT_NONE) at cdbdisp.c:133
#13 0x000055eaf846fc73 in CdbDispatchDtxProtocolCommand (dtxProtocolCommand=dtxProtocolCommand@entry=DTX_PROTOCOL_COMMAND_ABORT_NO_PREPARED, 
    dtxProtocolCommandLoggingStr=dtxProtocolCommandLoggingStr@entry=0x55eaf8984800 "Distributed Abort (No Prepared)", gid=gid@entry=0x7ffdd95418a0 "1677520340-0000964517", 
    qeError=qeError@entry=0x7ffdd9541840, numresults=numresults@entry=0x7ffdd954183c, badGangs=badGangs@entry=0x0, dtxSegments=0x55eaf9874e30, serializedDtxContextInfo=0x0, 
    serializedDtxContextInfoLen=0) at cdbdisp_dtx.c:117
#14 0x000055eaf84aa808 in doDispatchDtxProtocolCommand (dtxProtocolCommand=dtxProtocolCommand@entry=DTX_PROTOCOL_COMMAND_ABORT_NO_PREPARED, gid=gid@entry=0x7ffdd95418a0 "1677520340-0000964517", 
    badGangs=badGangs@entry=0x0, raiseError=raiseError@entry=0 '\000', dtxSegments=0x55eaf9874e30, serializedDtxContextInfo=serializedDtxContextInfo@entry=0x0, serializedDtxContextInfoLen=0)
    at cdbtm.c:1312
#15 0x000055eaf84aac2b in currentDtxDispatchProtocolCommand (dtxProtocolCommand=dtxProtocolCommand@entry=DTX_PROTOCOL_COMMAND_ABORT_NO_PREPARED, raiseError=raiseError@entry=0 '\000') at cdbtm.c:1273
#16 0x000055eaf84aba87 in doNotifyingAbort () at cdbtm.c:778
#17 0x000055eaf84abff5 in rollbackDtxTransaction () at cdbtm.c:1071
#18 0x000055eaf7fd7a48 in AbortTransaction () at xact.c:3409
#19 0x000055eaf7fd9ad5 in AbortCurrentTransaction () at xact.c:3927
#20 0x000055eaf82d39ee in PostgresMain (argc=1, argv=argv@entry=0x55eaf986aa40, dbname=<optimized out>, username=0x55eaf98063f8 "spt_main_user") at postgres.c:5053
#21 0x000055eaf7f654d2 in BackendRun (port=0x55eaf986e0f0) at postmaster.c:4843
#22 BackendStartup (port=0x55eaf986e0f0) at postmaster.c:4507
#23 ServerLoop () at postmaster.c:1956
#24 0x000055eaf826aa27 in PostmasterMain (argc=6, argv=0x55eaf9804200) at postmaster.c:1526
#25 0x000055eaf7f66ee7 in main (argc=6, argv=0x55eaf9804200) at main.c:245
```

The problem is NULL value for line variable at [this](https://github.com/greenplum-db/gpdb/blob/2d92fc4b5dbadaa643dc904a00cd96eb9274530e/src/backend/cdb/dispatcher/cdbconn.c#L717) line. See `src` below:
```
(gdb) frame 3
#3  0x000055eaf87ccaec in strlcpy (dst=dst@entry=0x55eaf9864888 "", src=src@entry=0x0, siz=siz@entry=10) at strlcpy.c:56
```

It appears like at least one of the segments gets OOM and coordinator tries to abort the transaction:
```
(gdb) print conn->errorMessage 
$1 = {data = 0x55eaf9887f60 "ERROR:  Out of memory\nDETAIL:  System memory limit reached, failed to allocate 524344 bytes from system\n", len = 104, maxlen = 256}
```

At `pqInternalNotice` the `res` contents is like that:
```
(gdb) print *res->errFields 
$3 = {next = 0x55eaf983b678, code = 83 'S', contents = "N"}
(gdb) print *res->errFields->next
$4 = {next = 0x0, code = 77 'M', contents = "m"}
```

So it appears, that [MPPnoticeReceiver](https://github.com/greenplum-db/gpdb/blob/2d92fc4b5dbadaa643dc904a00cd96eb9274530e/src/backend/cdb/dispatcher/cdbconn.c#L547) has no way of filling the `line` wariable.

My questions about MPPnoticeReceiver function are:
1) Why are we sure that `line` always gets a value? From what I see it actually rarely does, especially for notices.
2) Could we just make it `=""` as we do for other variables [here](https://github.com/greenplum-db/gpdb/blob/2d92fc4b5dbadaa643dc904a00cd96eb9274530e/src/backend/cdb/dispatcher/cdbconn.c#L554)?
3) Why cant we use COPY_VARLEN_FIELD macro [here](https://github.com/greenplum-db/gpdb/blob/2d92fc4b5dbadaa643dc904a00cd96eb9274530e/src/backend/cdb/dispatcher/cdbconn.c#L717)?

I'd say as a solution we could just initialize line as an empty string.

### Greenplum version or build
6.22.2

### OS version and uname -a
Ubuntu 18.04

### Expected behavior
Don't crash

### Actual behavior
Crashes

### Step to reproduce the behavior
Could not reproduce

**Comment by kainwen** [comment]
@Smyatkin-Maxim looks like the same as https://github.com/greenplum-db/gpdb/issues/14524

Not confirm yet.

**Comment by Smyatkin-Maxim** [comment]
Yep, seems like the same problem. Though for us the cause was (most likely) OOM on segment(s). No problems with resource groups as far as I can tell.

**Comment by ashwinstar** [comment]
@Smyatkin-Maxim your proposed fix seems fine wish to send PR?

**Comment by Smyatkin-Maxim** [comment]
Sure

**Comment by zxuejing** [comment]
hi @Smyatkin-Maxim, Does the core file still exist? if it exists, can you help to print res->errMsg in pqInternalNotice, thanks a lot!

**Comment by Smyatkin-Maxim** [comment]
Sure @zxuejing:
```
(gdb) print res->errMsg
$1 = 0x55eaf983b6ce "message type 0x43 arrived from server while idle\n"
```

**Comment by zxuejing** [comment]
hi @Smyatkin-Maxim  
Because we got an unexpected msg while idle So, we constructed a simple notice msg which do not specify the line no.
Although we get notice from QE in send_message_to_frontend, In extreme cases the line no may be 0.
If line no is 0, QE does not send the line no to QD. see the following codes.

```
if (edata->lineno > 0)
{
      snprintf(tbuf, sizeof(tbuf), "%d", edata->lineno);
      pq_sendbyte(&msgbuf, PG_DIAG_SOURCE_LINE);
      err_sendstring(&msgbuf, tbuf);
}
```
Based on the above two cases, we can infer that in MPPnoticeReceiver line no may be NULL. And this is reasonable.

After we reconstruct received notice msg in MPPnoticeReceiver, Then send the constructed notice message to client through forwardQENotices.

In forwardQENotices we can see if notice->line is null we do not send the line no info to the client.
```
if (notice->line)
{
	pq_sendbyte(&msgbuf,PG_DIAG_SOURCE_LINE);
	pq_sendstring(&msgbuf, notice->line);
}
```

So I think if line no is null, we do not copy it, and it will not be sent to the client.
```
if (line)
{
    strlcpy(notice->line, line, sizeof(notice->line));
}
```
I think if we do not have line no, do not send it to client is better than send 0 to client.
